File: abc201-220/abc204/e/main.py
from heapq import heappush, heappop

INF = 1 << 60


def isqrt(n: int) -> int:
	l, r = -1, n + 1
	while r - l > 1:
		m = (l + r) // 2
		if m * m <= n:
			l = m
		else:
			r = m
	return l


N, M = map(int, input().split())
edges = [[] for _ in range(N + 1)]
for i in range(M):
	a, b, c, d = map(int, input().split())
	edges[a].append((b, c, d))
	edges[b].append((a, c, d))

dp = [INF] * (N + 1)
dp[1] = 0
h = [(0, 1)]
while h:
	t, fr = heappop(h)
	if t > dp[fr]:
		continue
	for to, c, d in edges[fr]:
		s = t + c + d // (t + 1)
		rd = isqrt(d)
		for i in range(max(t, rd - 1), rd + 2):
			s = min(s, i + c + d // (i + 1))
		if s < dp[to]:
			dp[to] = s
			heappush(h, (s, to))

print(-1 if dp[N] == INF else dp[N])

File: abc201-220/abc204/e/main.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

const ll INF = 1LL << 60;

ll isqrt(ll n) {
	ll l = -1, r = n + 1;
	while (r - l > 1) {
		ll m = (l + r) / 2;
		if (m * m <= n) {
			l = m;
		} else {
			r = m;
		}

	}
	return l;

}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	int N, M;
	cin >> N >> M;
	vector<vector<tuple<int, ll, ll>>> edges(N + 1);
	for (int i = 0; i < M; i++) {
		int a, b;
		ll c, d;
		cin >> a >> b >> c >> d;
		edges[a].emplace_back(b, c, d);
		edges[b].emplace_back(a, c, d);
	}

	vector<ll> dp(N + 1, INF);
	dp[1] = 0;
	priority_queue<pair<ll, int>, vector<pair<ll, int>>, greater<pair<ll, int>>> q;
	q.emplace(0, 1);
	while (q.size()) {
		auto [t, from] = q.top();
		q.pop();
		if (t > dp[from]) {
			continue;
		}
		for (auto &&[to, c, d] : edges[from]) {
			ll s = t + c + d / (t + 1);
			ll rd = isqrt(d);
			for (ll i = max(t, rd - 1); i < rd + 1; i++) {
				s = min(s, i + c + d / (i + 1));
			}
			if (s < dp[to]) {
				dp[to] = s;
				q.emplace(s, to);
			}
		}

	}

	cout << (dp[N] == INF ? -1 : dp[N]) << "\n";

	return 0;
}


